### ISSUE TITLE
NaN in labels NDArray should represent irrelevant value, meaning no gradient and no loss.

### ISSUE THREAD

**Issue by strelec** [eclipse/deeplearning4j#4832]
Currently, there is no easy way to tell the training process that you just do not care about certain outputs of the neural network.

Expected:
[0.3, 1, 0, 1] - neural network output
[NaN, 1, 0, NaN] - labels
^ zero loss, perfect match

Actual:
NaN loss.

**Comment by treo** [comment]
Actually there is: Masking. Instead of conflating Labels with Non-Label information, you instead use masking to set what is relevant (1) or not relevant (0).

**Comment by strelec** [comment]
1. Masking does not work with softmax activation.
2. Because mask is applied after the activation function is computed (I have checked the source code), it is not apropriate here. Because big values that are masked, alerady affected the output.

**Comment by eraly** [comment]
Using the softmax with masking like that would not give the result you intend. Wondering if you don't care about certain outputs and you are training for classification (assumed based on softmax) why wouldn't you use sigmoid with binary cross entropy? What would work would be to replace the NaN with zeros in the labels and mask the way Treo says but without using softmax.

**Comment by treo** [comment]
> Because mask is applied after the activation function is computed (I have checked the source code), it is not apropriate here.

That's not quite right. What you see is that masking is applied after calculating the individual loss, the overall example loss is calculated at a later point: https://github.com/deeplearning4j/nd4j/blob/master/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMCXENT.java#L113-L126

And as the gradient is calculated with respect to each output, masking also works as I suggested:
https://github.com/deeplearning4j/nd4j/blob/master/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMCXENT.java#L185

Depending on the loss function it may be that masking works differently, but that is usually due to how the loss function works, and introducing NaNs into labels would in any case be a mistake.
